Iterate two or more objects over a joined table

Hi, thank for your work on this!

I wonder if I can assoicated an object by an id to another corresponding object during iteration.
i.e. Getting the object A, B during the iteration by inner join.
```
// SELECT A.*, B.*
// FROM A
// INNER JOIN B
// ON A.b_uid = B.uid
for (auto& [a, b] : storage.iterate<A, B>(inner_join<B>(on(c(&A::b_uid) == &B::uid)))) {
  // do something to a, b
  ...
}
```

According to the inner join [example](https://github.com/fnc12/sqlite_orm/blob/master/examples/left_and_inner_join.cpp),  now I can only use `select` with a batch size to avoid memory issue when the tables are too large to be loaded once.

`get_all` return objects but only for the left one. i.e. `get_all<A>(statement)`, so if there is a usage like `get_all<A, B>(statement)`, it'll also solve the problem by setting a reading batch size

Is there any hidden method I ignored to do that?

Hi @trueqbit , I used the header `include/sqlite_orm/sqlite_orm.h` in the branch `dev` and I tried to compile a tiny example but got an error.
```
#include "sqlite_orm.h"
#include <string>

struct A {
    int uid;
    int b_uid;
};

struct B {
    int uid;
};

int main() {
    using namespace sqlite_orm;
    auto storage = make_storage(
        "test.db",
        make_table("A",
                    make_column("uid", &A::uid),
                    make_column("b_uid", &A::b_uid)),
        make_table("B",
                    make_column("uid", &B::uid))
    );
    for (auto [a, b] : storage.iterate(
        select(
            columns(
                object<A>(true),
                object<B>(true)),
            inner_join<B>(on(c(&A::b_uid) == &B::uid))))) {
    // do something to a, b
    // ...
    }
    return 0;
}
```

error
```
test.cpp:28:56: error: no matching function for call to ‘sqlite_orm::internal::storage_t<sqlite_orm::internal::table_t<A, false, sqlite_orm::internal::column_t<int A::*, sqlite_orm::internal::empty_setter>, sqlite_orm::internal::column_t<int A::*, sqlite_orm::internal::empty_setter> >, sqlite_orm::internal::table_t<B, false, sqlite_orm::internal::column_t<int B::*, sqlite_orm::internal::empty_setter> > >::iterate(sqlite_orm::internal::select_t<sqlite_orm::internal::columns_t<sqlite_orm::internal::object_t<A>, sqlite_orm::internal::object_t<B> >, sqlite_orm::internal::inner_join_t<B, sqlite_orm::internal::on_t<sqlite_orm::internal::is_equal_t<int A::*, int B::*> > > >)’
   28 |             inner_join<B>(on(c(&A::b_uid) == &B::uid))))) {
      |                                                        ^
In file included from test.cpp:1:
sqlite_orm.h:23383:48: note: candidate: ‘template<class T, class O, class ... Args> sqlite_orm::internal::mapped_view<O, sqlite_orm::internal::storage_t<DBO>, Args ...> sqlite_orm::internal::storage_t<DBO>::iterate(Args&& ...) [with T = T; O = O; Args = {Args ...}; DBO = {sqlite_orm::internal::table_t<A, false, sqlite_orm::internal::column_t<int A::*, sqlite_orm::internal::empty_setter>, sqlite_orm::internal::column_t<int A::*, sqlite_orm::internal::empty_setter> >, sqlite_orm::internal::table_t<B, false, sqlite_orm::internal::column_t<int B::*, sqlite_orm::internal::empty_setter> >}]’
23383 |             mapped_view<O, self_type, Args...> iterate(Args&&... args) {
      |                                                ^~~~~~~
sqlite_orm.h:23383:48: note:   template argument deduction/substitution failed:
test.cpp:28:56: note:   couldn’t deduce template parameter ‘T’
   28 |             inner_join<B>(on(c(&A::b_uid) == &B::uid))))) {
      |                                                        ^
```
Did I miss anything?

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -27,8 +27,8 @@ environment:
 
     - job_name: clang, C++23
       appveyor_build_worker_image: Ubuntu2204
-      CC: clang-18
-      CXX: clang++-18
+      CC: clang-20
+      CXX: clang++-20
       SQLITE_ORM_CXX_STANDARD: "-DSQLITE_ORM_ENABLE_CXX_23=ON"
       cmake_build_parallel: ""
 

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -108,7 +108,7 @@ for:
   install:
     - |-
       cd C:\Tools\vcpkg
-      git fetch --tags && git checkout 2025.04.09
+      git fetch --tags && git checkout 2025.08.27
       cd %APPVEYOR_BUILD_FOLDER%
       C:\Tools\vcpkg\bootstrap-vcpkg.bat -disableMetrics
       C:\Tools\vcpkg\vcpkg integrate install
@@ -142,7 +142,7 @@ for:
   install:
     - |-
       pushd $HOME/vcpkg
-      git fetch --tags && git checkout 2025.04.09
+      git fetch --tags && git checkout 2025.08.27
       popd
       $HOME/vcpkg/bootstrap-vcpkg.sh -disableMetrics
       $HOME/vcpkg/vcpkg integrate install --overlay-triplets=vcpkg/triplets
@@ -170,7 +170,7 @@ for:
   # using custom vcpkg triplets for building and linking dynamic dependent libraries
   install:
     - |-
-      git clone --depth 1 --branch 2025.04.09 https://github.com/microsoft/vcpkg.git $HOME/vcpkg
+      git clone --depth 1 --branch 2025.08.27 https://github.com/microsoft/vcpkg.git $HOME/vcpkg
       $HOME/vcpkg/bootstrap-vcpkg.sh -disableMetrics
       $HOME/vcpkg/vcpkg integrate install --overlay-triplets=vcpkg/triplets
       vcpkg install sqlite3[core,dbstat,math,json1,fts5,soundex] catch2 --overlay-triplets=vcpkg/triplets

--- a/dev/result_set_iterator.h
+++ b/dev/result_set_iterator.h
@@ -12,7 +12,6 @@
 #include "column_result_proxy.h"
 #include "util.h"
 
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
 namespace sqlite_orm::internal {
 
     template<class ColResult, class DBOs>
@@ -71,6 +70,11 @@ namespace sqlite_orm::internal {
         friend bool operator==(const result_set_iterator& it, const result_set_sentinel_t&) noexcept {
             return sqlite3_data_count(it.stmt.get()) == 0;
         }
+#ifndef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
+        friend bool operator!=(const result_set_iterator& it, const result_set_sentinel_t& s) noexcept {
+            return !(it == s);
+        }
+#endif
 
       private:
         void step() {
@@ -87,4 +91,3 @@ namespace sqlite_orm::internal {
         statement_finalizer stmt;
     };
 }
-#endif

--- a/dev/result_set_view.h
+++ b/dev/result_set_view.h
@@ -4,7 +4,7 @@
 #ifndef SQLITE_ORM_IMPORT_STD_MODULE
 #include <utility>  //  std::move, std::remove_cvref
 #include <functional>  //  std::reference_wrapper
-#if defined(SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED) && defined(SQLITE_ORM_CPP20_RANGES_SUPPORTED)
+#ifdef SQLITE_ORM_CPP20_RANGES_SUPPORTED
 #include <ranges>  //  std::ranges::view_interface
 #endif
 #endif
@@ -18,7 +18,6 @@
 #include "type_traits.h"
 #include "storage_lookup.h"
 
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
 namespace sqlite_orm::internal {
     /*  
      *  A C++ view over a result set of a select statement, returned by `storage_t::iterate()`.
@@ -47,7 +46,7 @@ namespace sqlite_orm::internal {
 
         auto begin() {
             const auto& exprDBOs = db_objects_for_expression(this->db_objects.get(), this->expression);
-            using ExprDBOs = std::remove_cvref_t<decltype(exprDBOs)>;
+            using ExprDBOs = polyfill::remove_cvref_t<decltype(exprDBOs)>;
             // note: Select can be `select_t` or `with_t`
             using select_type = polyfill::detected_or_t<expression_type, expression_type_t, expression_type>;
             using column_result_type = column_result_of_t<ExprDBOs, select_type>;
@@ -81,5 +80,4 @@ namespace sqlite_orm::internal {
 #ifdef SQLITE_ORM_CPP20_RANGES_SUPPORTED
 template<class Select, class DBOs>
 inline constexpr bool std::ranges::enable_borrowed_range<sqlite_orm::internal::result_set_view<Select, DBOs>> = true;
-#endif
 #endif

--- a/dev/storage.h
+++ b/dev/storage.h
@@ -298,7 +298,6 @@ namespace sqlite_orm {
             }
 #endif
 
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
             /*  
              *  Iterate over a result set of a select statement.
              *  
@@ -330,7 +329,6 @@ namespace sqlite_orm {
                 auto connection = this->get_connection();
                 return {this->db_objects, std::move(connection), std::move(expression)};
             }
-#endif
 #endif
 
 #ifdef SQLITE_ORM_CPP23_GENERATOR_SUPPORTED

--- a/examples/iteration.cpp
+++ b/examples/iteration.cpp
@@ -66,14 +66,12 @@ int main(int, char**) {
     }
     cout << "heroesByAlgorithm.size = " << heroesByAlgorithm.size() << endl;
 
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
     cout << "====" << endl;
 
     cout << "Distinct hero names:" << endl;
     for (std::string name: storage.iterate(select(distinct(&MarvelHero::name)))) {
         cout << name << endl;
     }
-#endif
 
     return 0;
 }

--- a/include/sqlite_orm/sqlite_orm.h
+++ b/include/sqlite_orm/sqlite_orm.h
@@ -16414,7 +16414,7 @@ inline constexpr bool std::ranges::enable_borrowed_range<sqlite_orm::internal::m
 #ifndef SQLITE_ORM_IMPORT_STD_MODULE
 #include <utility>  //  std::move, std::remove_cvref
 #include <functional>  //  std::reference_wrapper
-#if defined(SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED) && defined(SQLITE_ORM_CPP20_RANGES_SUPPORTED)
+#ifdef SQLITE_ORM_CPP20_RANGES_SUPPORTED
 #include <ranges>  //  std::ranges::view_interface
 #endif
 #endif
@@ -16440,7 +16440,6 @@ inline constexpr bool std::ranges::enable_borrowed_range<sqlite_orm::internal::m
 
 // #include "util.h"
 
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
 namespace sqlite_orm::internal {
 
     template<class ColResult, class DBOs>
@@ -16499,6 +16498,11 @@ namespace sqlite_orm::internal {
         friend bool operator==(const result_set_iterator& it, const result_set_sentinel_t&) noexcept {
             return sqlite3_data_count(it.stmt.get()) == 0;
         }
+#ifndef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
+        friend bool operator!=(const result_set_iterator& it, const result_set_sentinel_t& s) noexcept {
+            return !(it == s);
+        }
+#endif
 
       private:
         void step() {
@@ -16515,7 +16519,6 @@ namespace sqlite_orm::internal {
         statement_finalizer stmt;
     };
 }
-#endif
 
 // #include "ast_iterator.h"
 
@@ -16527,7 +16530,6 @@ namespace sqlite_orm::internal {
 
 // #include "storage_lookup.h"
 
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
 namespace sqlite_orm::internal {
     /*  
      *  A C++ view over a result set of a select statement, returned by `storage_t::iterate()`.
@@ -16556,7 +16558,7 @@ namespace sqlite_orm::internal {
 
         auto begin() {
             const auto& exprDBOs = db_objects_for_expression(this->db_objects.get(), this->expression);
-            using ExprDBOs = std::remove_cvref_t<decltype(exprDBOs)>;
+            using ExprDBOs = polyfill::remove_cvref_t<decltype(exprDBOs)>;
             // note: Select can be `select_t` or `with_t`
             using select_type = polyfill::detected_or_t<expression_type, expression_type_t, expression_type>;
             using column_result_type = column_result_of_t<ExprDBOs, select_type>;
@@ -16590,7 +16592,6 @@ namespace sqlite_orm::internal {
 #ifdef SQLITE_ORM_CPP20_RANGES_SUPPORTED
 template<class Select, class DBOs>
 inline constexpr bool std::ranges::enable_borrowed_range<sqlite_orm::internal::result_set_view<Select, DBOs>> = true;
-#endif
 #endif
 
 // #include "ast_iterator.h"
@@ -23411,7 +23412,6 @@ namespace sqlite_orm {
             }
 #endif
 
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
             /*  
              *  Iterate over a result set of a select statement.
              *  
@@ -23443,7 +23443,6 @@ namespace sqlite_orm {
                 auto connection = this->get_connection();
                 return {this->db_objects, std::move(connection), std::move(expression)};
             }
-#endif
 #endif
 
 #ifdef SQLITE_ORM_CPP23_GENERATOR_SUPPORTED

--- a/tests/iterate.cpp
+++ b/tests/iterate.cpp
@@ -61,13 +61,18 @@ TEST_CASE("Iterate mapped") {
     }
 }
 
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
 TEST_CASE("Iterate select statement") {
     struct Test {
         int64_t id = 0;
         std::vector<char> key;
 
+#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
         bool operator==(const Test&) const = default;
+#else
+        bool operator==(const Test& right) const {
+            return this->id == right.id && this->key == right.key;
+        }
+#endif
     };
 #ifdef SQLITE_ORM_CPP20_CONCEPTS_SUPPORTED
     constexpr orm_table_reference auto test_table = c<Test>();
@@ -144,4 +149,3 @@ TEST_CASE("Iterate select statement") {
 #endif
 #endif
 }
-#endif

--- a/tests/static_tests/iterator_t.cpp
+++ b/tests/static_tests/iterator_t.cpp
@@ -9,12 +9,10 @@
 using namespace sqlite_orm;
 using internal::mapped_iterator;
 using internal::mapped_view;
-using internal::structure;
-#ifdef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
 using internal::result_set_iterator;
 using internal::result_set_sentinel_t;
 using internal::result_set_view;
-#endif
+using internal::structure;
 #ifdef SQLITE_ORM_WITH_CPP20_ALIASES
 using internal::table_reference;
 #endif
@@ -81,7 +79,7 @@ concept storage_iterate_mapped_ref = requires(S& storage_type) {
 };
 #endif
 
-#if defined(SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED) && defined(SQLITE_ORM_CPP20_CONCEPTS_SUPPORTED)
+#ifdef SQLITE_ORM_CPP20_CONCEPTS_SUPPORTED
 template<class Iter, class Value>
 concept can_iterate_result_set = requires(Iter it) {
     requires std::input_iterator<Iter>;
@@ -190,7 +188,7 @@ TEST_CASE("can view and iterate mapped") {
 #endif
 }
 
-#if defined(SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED) && defined(SQLITE_ORM_CPP20_CONCEPTS_SUPPORTED)
+#ifdef SQLITE_ORM_CPP20_CONCEPTS_SUPPORTED
 TEST_CASE("can view and iterate result set") {
     struct Object {};
     using empty_storage_type = decltype(make_storage(""));

